### PR TITLE
feat: add an OscOutput operator and its example

### PR DESCRIPTION
**Pull request by ylmrx**
This is a new operator, counterpart to OscInput
It _only_ supports strings and float numbers, it tries to fail gracefully on bad location and failure to connect.
Error and state check is grunty, please advise if I'm doing it too wrong.

### CODE REVIEW

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+using System.Net;
+using Rug.Osc;
+using System;
+using T3.Core.Utils;
+
+namespace T3.Operators.Types.Id_4e99da86_482f_4037_8664_b2371526d632
+{
+    public class OscOutput : Instance<OscOutput>
+    {
+        private bool _connected;
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:14]: Please always move private members to the end for the class.

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        private IPAddress _address;
+        private int _port;
+        private float _num_value;
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:17]:
```suggestion
        private float _numericValue;
```
Please follow the .net c# naming conventions and avoid short forms.

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        private string _str_value;
+        private OscSender _sender;
+        public OscOutput()
+        {
+            Result.UpdateAction = Update;
+        }
+
+        private enum OscType
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:25]:
```suggestion
        private enum OscTypes
```

Please move to bottom and use plural for enumerations.

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        {
+            Number,
+            String
+        }
+        private void Update(EvaluationContext context)
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:30]: Please leave blank line after closing `}`

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                }
+            }
+            catch (Exception e) {
+                Log.Error("Connection failure: " + e);
+            }
+            _connected = false;
+        }
+
+        [Input(Guid = "98f38caa-4f79-425c-ab46-22d7dbe62978")]
+        public readonly InputSlot<string> Address = new InputSlot<string>();
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:121]: Would `IpAddress` be a better name?

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            String
+        }
+        private void Update(EvaluationContext context)
+        {
+            var location = Location.GetValue(context);
+            var num = Number.GetValue(context);
+            var str = String.GetValue(context);
+            var oscType = OscValueType.GetEnumValue<OscType>(context);
+
+            CheckAddr(context);
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:37]:
```suggestion
            if(!TryGetValidAdress(IdAddress.GetValue(context), out var problem, out var address) {
               _lastErrorMessage = problem;
            }
```


I would have expected a `CheckXYZ` method to return a boolean. 
I much prefer the `TryGetXYZ` pattern being as static as possible.

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                    if (num != _num_value)
+                    {
+                        // Log.Info("Sending:" + num);
+                        OscMessage message = new OscMessage(location, num);
+                        if (message.Error == OscPacketError.None)
+                        {
+                            _sender.Send(message);
+                            _num_value = num;
+                        }
+                        else Log.Error("failed to build the OSC message (float): " + message.ErrorMessage);
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:55]: Please start messages with an uppercase letter.

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+namespace T3.Operators.Types.Id_4e99da86_482f_4037_8664_b2371526d632
+{
+    public class OscOutput : Instance<OscOutput>
+    {
+        private bool _connected;
+        private IPAddress _address;
+        private int _port;
+        private float _num_value;
+        private string _str_value;
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:18]:
```suggestion
        private string _stringValue;
```

File: Operators/Types/lib/io/osc/OscOutput.cs [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        [Input(Guid = "d5e9e9be-093b-4d57-9070-9b2cf33aa45b")]
+        public readonly InputSlot<string> String = new InputSlot<string>();
+
+        [Input(Guid = "827c7ae2-c129-4d04-a715-328c0a86bf8a")]
+        public readonly InputSlot<float> Number = new InputSlot<float>();
+
+        [Input(Guid = "4931217b-e30d-42a6-9cda-bc49a8fd8d67", MappedType = typeof(OscType))]
+        public readonly InputSlot<int> OscValueType = new ();
+
+        [Output(Guid = "a6679d6c-fc34-4588-ab20-5079ad8f8a03")]
```
Comment (pixtur) [inline Operators/Types/lib/io/osc/OscOutput.cs:139]: Outputs should be at the beginning of the file. (I know... it's strange. It's not my convention. But at least it is one).

**Review by pixtur — CHANGES_REQUESTED** [review]
That's an excellent PR! Most of my comments are nitpicky about code-style and conventions. But there are three things that should be done:

1. The Op should implement the Dispose-Pattern and probably disconnect when disposing. (If not, it might happen that the op keeps an connection open).
2. Another problem might be that too many open connections will lead to performance issues. I'm not sure if this is true but OscInput uses a shared OscConnectionManager to pool and reusing connections. Although this optimization might be quite involved. I think we can live without that here.
3. Please implement `IStatusProvider`.  This will show a warning icon whenever the op has a problem (like a missing connection or an invalid IP address). Check [AbletonLiveSync] for an example.

**Comment by ylmrx** [comment]
(disregard latest commit, involving very experimental/wip/poc regarding text support for OscInput, i wanna keep this change focused on OscOutput), switching it to draft, as i might be noisy in between)

**Comment by ylmrx** [comment]
> That's an excellent PR! Most of my comments are nitpicky about code-style and conventions. But there are three things that should be done:

Thanks ! :)

>     1. The Op should implement the Dispose-Pattern and probably disconnect when disposing. (If not, it might happen that the op keeps an connection open).

done.

>     2. Another problem might be that too many open connections will lead to performance issues. I'm not sure if this is true but OscInput uses a shared OscConnectionManager to pool and reusing connections. Although this optimization might be quite involved. I think we can live without that here.

The conManager is built with a server purpose in mind, and won't be much use for us here.

I'm not sure a threaded model would help us in this context, as there's no error check on the `OscRug.foobar.Send()` (we're UDP), and a given operator should address its whole traffic (all distinct IPs/port/location)

We can check a later stage, for a threaded model, if we get performance issue, but i think we're fine here.

On another, perf-related note, the operator gives a tiny glitch once in while, but seems to be related to the refresh rate of Tooll (and phase with touchosc frame rate (where you can't disable vsync) : it completely disappear when I disable vsync (in tooll) and it stays well under 1ms frame-time).

This is a general issue with OSC client (TouchOSC does that too (this is why many standalone sequencer/drumbox were attempted and failed (15ms is barely noticeable for eye, but sound like a clumsy drummer for a sequencer)

>     3. Please implement `IStatusProvider`.  This will show a warning icon whenever the op has a problem (like a missing connection or an invalid IP address). Check [AbletonLiveSync] for an example.

Done!

**Comment by pixtur** [comment]
Awesome!

**Pull request merged**